remote source implementation for meals.

cesards opened the pull request:
## Description
Recipe remote data source implementation.

## Context
The repository's contributors need data to start working on their MVI implementations. Even relying on real data is not a blocker, this is going to make things easier for everyone so they don't have to come up with their own ways to mock the data.

## Final Checks
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

cesards commented:
<img width="961" alt="Screenshot 2021-01-25 at 10 07 41 AM" src="https://user-images.githubusercontent.com/1536079/105723792-2ffc6280-5ef5-11eb-9fcf-46e31217fbc8.png">

Awesome, tests are failing as expected :) CI works.

cesards commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RecipeServiceProvider.kt:
```diff
@@ -0,0 +1,17 @@
+
+internal class RecipeServiceProvider(private val retrofit: Retrofit = RetrofitProvider.value) {
+
+    private var singleton: RecipeService? = null
+    val value: RecipeService
+        get() {
+            if (singleton == null) {
+                synchronized(this) {
+                    singleton = retrofit.create(RecipeService::class.java)
+                }
+            }
+            return singleton!!
+        }
```
Manual singleton, in order to avoid DI frameworks.

cesards commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RetrofitProvider.kt:
```diff
@@ -0,0 +1,21 @@
+
+object RetrofitProvider {
+
+    private var singleton: Retrofit? = null
+    val value: Retrofit
+        get() {
+            if (singleton == null) {
+                synchronized(this) {
+                    singleton = Retrofit.Builder()
+                        .baseUrl("https://www.themealdb.com/api/json/v1/1")
+                        .addConverterFactory(MoshiConverterFactory.create())
+                        .build()
+                }
+            }
+            return singleton!!
+        }
```
Manual singleton, in order to avoid DI frameworks.

cesards commented on gradle/android-library-common.gradle:
```diff
@@ -15,9 +15,9 @@ android {
-  kotlinOptions {
-    freeCompilerArgs = ["-Xallow-result-return-type"]
+    freeCompilerArgs += [
+        "-Xopt-in=kotlinx.coroutines.ExperimentalCoroutinesApi",
+        "-Xallow-result-return-type",
```
So we can use kotlin's `Result` as output and coroutines without warnings :)

Mr-Gibbs replied:
Man oh man you just got alllll new new toys :P

cesards commented on .github/workflows/main-integration-checks.yml:
```diff
@@ -56,7 +56,7 @@ jobs:
 
       - name: Detekt
-        run: ./gradlew detekt
+        run: ./gradlew detektMain
```
Detekt was broken due to the latest version changes introduced by the framework.

cesards commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RecipeServiceProvider.kt:
```diff
@@ -0,0 +1,17 @@
+internal class RecipeServiceProvider(private val retrofit: Retrofit = RetrofitProvider.value) {
+
+    private var singleton: RecipeService? = null
+    val value: RecipeService
+        get() {
+            if (singleton == null) {
+                synchronized(this) {
+                    singleton = retrofit.create(RecipeService::class.java)
+                }
+            }
+            return requireNotNull(singleton)
+        }
```
Manual singleton, in order to avoid DI frameworks.

cesards commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RetrofitProvider.kt:
```diff
@@ -0,0 +1,21 @@
+object RetrofitProvider {
+
+    private var singleton: Retrofit? = null
+    val value: Retrofit
+        get() {
+            if (singleton == null) {
+                synchronized(this) {
+                    singleton = Retrofit.Builder()
+                        .baseUrl("https://www.themealdb.com/api/json/v1/1")
+                        .addConverterFactory(MoshiConverterFactory.create())
+                        .build()
+                }
+            }
+            return requireNotNull(singleton)
+        }
```
Manual singleton, in order to avoid DI frameworks.

Mr-Gibbs commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RetrofitProvider.kt:
```diff
@@ -0,0 +1,21 @@
+object RetrofitProvider {
+
+    private var singleton: Retrofit? = null
+    val value: Retrofit
```
should we do full double check locking?

cesards replied:
It's been a really long time since I heard about that technique. I could make changes but I guess there a few things to have in consideration:

1. Since we are always creating the dependency when initializing objects (at construction time), multithreading shouldn't be a big issue, should it?

2. I've read a few things about the pattern and it doesn't seem to be recommended as a general practice (in favor of other lazy init approaches). See for reference:

- https://www.baeldung.com/java-singleton-double-checked-locking
- https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

Could we get more context on why this would be needed?

dlemures replied:
There are alternatives, but double check locking is pretty good for a few situations.

I guess that what @Mr-Gibbs means is that we are trying to do it thread-safe by using `synchronized`, but actually it is not.

dlemures replied:
If we wanna use lazy, we can use lazy {} ?

AlessandroOddone replied:
+1 for `lazy` or even just relying on the lazy initialization of `object`s and directly go:

```
val value: Retrofit = Retrofit.Builder()
  .baseUrl("https://www.themealdb.com/api/json/v1/1")
  .addConverterFactory(MoshiConverterFactory.create())
  .build()
```

cesards replied:
> If we wanna use lazy, we can use lazy {} ?

Both look good to me! I had actually thought about using `lazy` in the first place. I tried to create our own delegate but it was too verbose for what we needed. Thanks for bringing this up guys!

Actually, I looked at the implementation of `lazy` again (it's been a while), and I saw they actually used the full double check locking pattern you were talking about, right?

```
...
if (_v1 !== UNINITIALIZED_VALUE) {
      @Suppress("UNCHECKED_CAST")
      return _v1 as T
  }

  return synchronized(lock) {
      val _v2 = _value
      if (_v2 !== UNINITIALIZED_VALUE) {
          @Suppress("UNCHECKED_CAST") (_v2 as T)
          ...
  ...
```

cesards replied:
Addressed: [`4606ac9` (#13)](https://github.com/CodeForVictory/recipes-mvi-playground/pull/13/commits/4606ac9ce1e6d7155cba6ed2f1f3145ed6ee3929)

Mr-Gibbs commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/AdaptToMeal.kt:
```diff
@@ -0,0 +1,17 @@
+import org.codeforvictory.android.recipes.data.api.recipe.model.ApiMeal
+import org.codeforvictory.recipes.domain.common.model.meal.Meal
+
+class AdaptToMeal {
```
I'd personally prefer the verb Convert.  Also, a pattern I've used in the past that I've liked is having converters specified as extension functions on the Network type.  That lets you name it something like 

```
fun ApiModel.toDomain() : Model 
```

thoughts?

cesards replied:
The only potential issue I see when using extension functions with anemic modules is the testing side of things. If for any reason we need to mock `ApiModel`, we might encounter a few gotchas when calling `toDomain()`. Does it make sense?

AlessandroOddone replied:
I also prefer to have a top-level extension function (`ApiMeal.toMeal()`) to do this conversion. IMO not being able to mock the API models or the conversion logic itself is a feature in the sense that it helps ensure that non-trivial logic doesn't leak into these components.

cesards replied:
How would you scale adapters/converters if the model had more complex information? The first example that comes to mind is time conversion. If we depend on some other domain model to adjust to commonly expected data formatting, we would have something like:
```
class AdaptToMeal(private clock: Clock) {
	operator fun invoke(apiMeal: ApiMeal): Meal {
        return with(apiMeal) {
            Meal(
                id = id,
                name = name,
                imageThumbUrl = imageThumbnailUrl,
				currentTime = clock.currentTime()
            )
        }
    }
}
```
How would we tackle this problem by using top-level extension functions?

AlessandroOddone replied:
My assumption is that I would rarely need to do something complex in this layer since I expect most of the conversions to be presentation-related. So, I would only introduce an adapter when the more complex use case arises and not up front. This argument entirely depends on how the network responses look like in practice, so I'm fine defaulting to this pattern in this repo.

Mr-Gibbs commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RecipeService.kt:
```diff
@@ -0,0 +1,11 @@
+interface RecipeService {
+
+    @GET
+    suspend fun filter(@Query("a") area: String = "American"): ApiMealsResponse
```
IMO naming this filter is odd.  I'd usually see it named after the return (getRecipies etc).  The filtering bit is an implementation detail of the passed query param

cesards replied:
Sure. I actually thought a bit about the naming but I didn't have a better suggestion at the time. I think `recipesBy(...)` would make sense since we always need to specify a filter no matter what (although I provided a default argument because otherwise the API doesn't work (it's a tradeoff of having to deal with it).

dlemures replied:
Usually i see the patter `getX`.

cesards replied:
`get and set` names usually come from the Java world. As discussed offline with Daniel, we decided to go for `recipesBy(...)`. Do you like it better @Mr-Gibbs ?

Mr-Gibbs commented on app-api-common/recipe/src/test/java/org/codeforvictory/android/recipes/data/api/recipe/AdaptToMealTest.kt:
```diff
@@ -0,0 +1,24 @@
+    private val adaptToMeal = AdaptToMeal()
+
+    @Test
+    fun `invoke - should convert api into domain meal data`() {
```
man I just can't get used to these backticked function names with spaces :)

cesards replied:
I think it's easier to read natural language than having to deal with test cases like:

```
invoke_shouldDoThis_whenGivenXYZ
```

dlemures replied:
Cesar has my vote lol

AlessandroOddone replied:
Team backticks! 🥇

Mr-Gibbs reviewed (commented):
Few areas of discussion I think but otherwise looks good!

dlemures commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RecipeServiceProvider.kt:
```diff
@@ -0,0 +1,17 @@
+
+internal class RecipeServiceProvider(private val retrofit: Retrofit = RetrofitProvider.value) {
+
+    private var singleton: RecipeService? = null
```
The property is a new one every time we create a new RecipeServiceProvider, right? Shouldn't we make it equivalent to static to be a singleton?

dlemures replied:
Maybe object?

cesards replied:
Addressed: [`4606ac9` (#13)](https://github.com/CodeForVictory/recipes-mvi-playground/pull/13/commits/4606ac9ce1e6d7155cba6ed2f1f3145ed6ee3929)

dlemures commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RecipeServiceProvider.kt:
```diff
@@ -0,0 +1,17 @@
+        get() {
+            if (singleton == null) {
+                synchronized(this) {
+                    singleton = retrofit.create(RecipeService::class.java)
```
here we another check inside the synchronized, otherwise we will be create the singleton multiple times:
https://www.baeldung.com/java-singleton-double-checked-locking

cesards replied:
Addressed: [`4606ac9` (#13)](https://github.com/CodeForVictory/recipes-mvi-playground/pull/13/commits/4606ac9ce1e6d7155cba6ed2f1f3145ed6ee3929)

dlemures commented on app-api-common/recipe/src/test/java/org/codeforvictory/android/recipes/data/api/recipe/AdaptToMealTest.kt:
```diff
@@ -0,0 +1,24 @@
+        // Then
+        result.id shouldBeEqualTo apiMeal.id
+        result.name shouldBeEqualTo apiMeal.name
+        result.imageThumbUrl shouldBeEqualTo apiMeal.imageThumbnailUrl
```
love it

dlemures commented on app-domain-common/model/src/main/java/org/codeforvictory/recipes/domain/common/model/meal/Meal.kt:
```diff
@@ -0,0 +1,3 @@
+package org.codeforvictory.recipes.domain.common.model.meal
+
+class Meal(val id: String, val name: String, val imageThumbUrl: String)
```
this might be a stupid question. Not data class cause of native?

Mr-Gibbs replied:
I almost asked the same.  I'm assuming it's just a matter of don't make it data class unless you need the freebies it gives?

cesards replied:
This was a slip! Thank you both for catching it up. Addressed: [`4606ac9` (#13)](https://github.com/CodeForVictory/recipes-mvi-playground/pull/13/commits/4606ac9ce1e6d7155cba6ed2f1f3145ed6ee3929)

Mr-Gibbs commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RecipeApiSource.kt:
```diff
@@ -0,0 +1,17 @@
+import org.codeforvictory.recipes.domain.common.model.meal.Meal
+
+class RecipeApiSource(
+    private val adaptToMeal: AdaptToMeal = AdaptToMeal(),
```
It's weird to me to have to inject/instantiate what really amounts to a pure, static function (conversion).

cesards replied:
See my comment: https://github.com/CodeForVictory/recipes-mvi-playground/pull/13#discussion_r567547895

dlemures commented on app-api-common/recipe/src/main/java/org/codeforvictory/android/recipes/data/api/recipe/RecipeApiSource.kt:
```diff
@@ -0,0 +1,17 @@
+
+class RecipeApiSource(
+    private val adaptToMeal: AdaptToMeal = AdaptToMeal(),
+    private val recipeService: RecipeService = RecipeServiceProvider().value,
```
talking offline Cesar and I:

* Not using default arguments
* We do not need RecipeServiceProvider
* We can keep RetrofitProvider to have a single place where we create the retrofit config

cesards replied:
Addressed: https://github.com/CodeForVictory/recipes-mvi-playground/pull/13/commits/4606ac9ce1e6d7155cba6ed2f1f3145ed6ee3929

cesards commented:
@dlemures @Mr-Gibbs @AlessandroOddone Please have another look, I addressed most of the feedback. Feel free to mark conversations as resolved and/or 👍 the PR if it looks ok to you.

On Friday I had a call with Daniel and we thought it would be a good idea to ditch the Providers and default arguments for the data components to favor our own DI, so we choose how to create those dependencies. I added `javax.annotation` dependency in case any of us don't want to do manual injection, since it doesn't hurt.

Let me know if I missed something else or any other thing, really.

AlessandroOddone reviewed (approved):
LGTM 👍

dlemures commented:
LGTM
